### ISSUE TITLE
"Check.DebugAssert failed: Only call this method for properties known to be part of a key." on BulkSaveChangesAsync

### ISSUE THREAD

**Issue by schuettecarsten** [zzzprojects/EntityFramework-Extensions#451]
### Description
I receive the following exception when calling BulkSaveChangesAsync(). The change tracker contains a bunch of modifications, so it's very hard to find out which one is causing this exception. The code works with SaveChangesAsync(). As the ZZZ code is obfuscated, I cannot find out any details with the debugger.

### Exception
```
Check.DebugAssert failed: Only call this method for properties known to be part of a key.
   at Microsoft.EntityFrameworkCore.Utilities.Check.DebugAssert(Boolean condition, String message)
   at Microsoft.EntityFrameworkCore.Metadata.Internal.PropertyExtensions.IsForeignKeyToSelf(IReadOnlyProperty property)
   at Z.EntityFramework.Extensions.PropertyZInfo..ctor(IProperty property)
   at PublicExtensions..(IProperty )
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at PublicExtensions.ToZInfo(IProperty property)
   at Z.EntityFramework.Extensions.EntityTypeZInfo..ctor(IEntityType entityType)
   at PublicExtensions..(IEntityType )
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at PublicExtensions.ToZInfo(IEntityType entityType)
   at .BulkUpdate[T](DbContext this, IEnumerable`1 entities, Action`1 options, Boolean isBulkSaveChanges)
   at .(DbContext this, List`1 , Action`1 )
   at .(DbContext this, StateManager , IReadOnlyList`1 , Action`1 )
   at .(DbContext this, StateManager , IReadOnlyList`1 , Action`1 )
   at .(DbContext this, Action`1 , DbContext )
   at DbContextExtensions.BulkSaveChanges(DbContext this, Action`1 options)
   at DbContextExtensions..()
   at System.Threading.Tasks.Task.InnerInvoke()
   at System.Threading.Tasks.Task.<>c.<.cctor>b__271_0(Object obj)
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
--- End of stack trace from previous location ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable.ConfiguredTaskAwaiter.GetResult()
   at DbContextExtensions..MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable.ConfiguredTaskAwaiter.GetResult()
   at Neusta.Shared.DataAccess.EntityFramework.ExtendedEntityFrameworkDataContext`2.<InternalSaveChangesAsync>d__35.MoveNext() in D:\Projekte\Azure\CloudManagementTool\Source\SharedLibrary\Neusta.Shared.DataAccess.EntityFramework.Extensions\ExtendedEntityFrameworkDataContext.Generic.cs:line 1248
```

### Further technical details
- EF version: EF Core 6.0.0 from latest .NET 6.0 release
- EF Extensions version: 6.13.0 (latest available)
- Database Provider: SqlServer

**Comment by JonathanMagnan** [comment]
Hello @schuettecarsten ,

Do you think you could provide us with a runnable project with the minimum code to reproduce this issue? You can send it in private here if needed: info@zzzprojects.com

I would recommend you if possible to try first with a more recent version of Entity Framework such as the latest version 6.4.4, perhaps this issue was inside entity framework and is already fixed.

Best Regards,

Jon

**Comment by schuettecarsten** [comment]
Hi, we are talking about EF Core and the version that ships with latest .NET 6.0 release.

**Comment by JonathanMagnan** [comment]
Oh my bad, I will need to be more careful when seeing "6.0" now ;)

It is still possible for you to provide a project? There is something in your model that we do not support but we don't know how to reproduce it.

**Comment by JonathanMagnan** [comment]
Hello again!

Since our last conversation, we haven't heard from you!

Don't hesitate to provide a project so we can assist you better.

Looking forward to hearing from  you, 

Jon

**Comment by JonathanMagnan** [comment]
Hello again! 

A simple follow-up to see if you had time to provide a runnable project? 

Don't hesitate to contact us for further assistance. 

Best regards,

Jon

**Comment by schuettecarsten** [comment]
Unfortunately, I was not able to create a simple project, so the issue is caused by something that we do with the model (there are several custom conventions), but the way how ZZZ is obfuscated makes it impossible for me to dig deeper into this issue. Workaround is to use standard SaveChangesAsync for now.

**Comment by JonathanMagnan** [comment]
With the help of @schuettecarsten , this issue has been finally fixed in the v6.13.4

Thank @schuettecarsten for your contribution ;)
